File: nd4j/samediff-audio/src/main/java/org/eclipse/deeplearning4j/audio/feature/WhisperMelSpectrogram.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.eclipse.deeplearning4j.audio.feature;

import lombok.extern.slf4j.Slf4j;
import org.eclipse.deeplearning4j.audio.io.AudioLoader;
import org.eclipse.deeplearning4j.audio.transform.AudioPreprocessor;
import org.eclipse.deeplearning4j.audio.whisper.WhisperConfig;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.impl.audio.WhisperMelSpectrogramOp;
import org.nd4j.linalg.factory.Nd4j;

import java.io.File;
import java.io.IOException;

/**
 * Whisper-specific mel spectrogram extraction using native C++ op.
 * <p>
 * Delegates to the native {@code whisper_mel_spectrogram} op which combines
 * STFT + mel filterbank + Whisper-specific log normalization in a single kernel:
 * <ul>
 *   <li>16kHz sample rate, N_FFT=400, HOP_LENGTH=160</li>
 *   <li>80 or 128 mel bins (depending on model version)</li>
 *   <li>Pad/trim to 3000 frames (30-second chunks)</li>
 *   <li>log10(max(mel, 1e-10)), clamp to max-8.0, then (x + 4.0) / 4.0</li>
 * </ul>
 */
@Slf4j
public class WhisperMelSpectrogram {

    private final WhisperConfig config;

    public WhisperMelSpectrogram(WhisperConfig config) {
        this.config = config;
    }

    /**
     * Extract Whisper mel spectrogram features from raw audio samples.
     *
     * @param audio Audio samples as FLOAT INDArray of shape [samples], at 16kHz
     * @return Mel spectrogram of shape [1, numMelBins, 3000]
     */
    public INDArray extractFeatures(INDArray audio) {
        // Pad or trim to exactly 30 seconds (480000 samples at 16kHz)
        int chunkSamples = config.getChunkLengthSamples();
        INDArray padded = AudioPreprocessor.padOrTrim(audio, chunkSamples);

        // Ensure rank 2 for batch processing: [1, samples]
        if (padded.rank() == 1) {
            padded = padded.reshape(1, padded.length());
        }

        // Run native whisper_mel_spectrogram op (mel + log normalization in one call)
        INDArray result = Nd4j.exec(new WhisperMelSpectrogramOp(
                padded,
                config.getSampleRate(),
                config.getNFft(),
                config.getHopLength(),
                config.getNumMelBins(),
                config.getNumFrames(),
                0.0,   // lower edge Hz
                (double) config.getSampleRate() / 2.0  // upper edge Hz (Nyquist)
        ))[0];

        // Ensure shape is [1, numMelBins, numFrames]
        if (result.rank() == 2) {
            result = result.reshape(1, result.size(0), result.size(1));
        }

        return result.castTo(DataType.FLOAT);
    }

    /**
     * Extract features from a WAV file.
     *
     * @param wavFile Path to the WAV file
     * @return Mel spectrogram of shape [1, numMelBins, 3000]
     * @throws IOException if the file cannot be read
     */
    public INDArray extractFeaturesFromFile(File wavFile) throws IOException {
        INDArray audio = AudioLoader.loadWav(wavFile);
        int sourceSampleRate = AudioLoader.getSampleRate(wavFile);

        // Resample to 16kHz if needed
        if (sourceSampleRate != config.getSampleRate()) {
            AudioPreprocessor preprocessor = AudioPreprocessor.builder()
                    .targetSampleRate(config.getSampleRate())
                    .normalize(false)
                    .applyPreEmphasis(false)
                    .build();
            audio = preprocessor.process(audio, sourceSampleRate);
        }

        return extractFeatures(audio);
    }
}

File: libnd4j/include/ops/declarable/generic/audio/whisper_mel_spectrogram.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author Adam Gibson
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_whisper_mel_spectrogram)

#include <ops/declarable/headers/audio.h>
#include <ops/declarable/helpers/audio.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(whisper_mel_spectrogram, 1, 1, false, -2, -2) {
    auto input = INPUT_VARIABLE(0);
    auto output = OUTPUT_VARIABLE(0);

    int sampleRate = block.numI() > 0 ? INT_ARG(0) : 16000;
    int fftSize = block.numI() > 1 ? INT_ARG(1) : 400;
    int hopLength = block.numI() > 2 ? INT_ARG(2) : 160;
    int numMelBins = block.numI() > 3 ? INT_ARG(3) : 80;
    int targetFrames = block.numI() > 4 ? INT_ARG(4) : 3000;
    double lowerEdgeHz = block.numT() > 0 ? T_ARG(0) : 0.0;
    double upperEdgeHz = block.numT() > 1 ? T_ARG(1) : 8000.0;

    REQUIRE_TRUE(input->rankOf() == 1 || input->rankOf() == 2, 0,
                 "whisper_mel_spectrogram: input must be rank 1 or 2, got %d", input->rankOf());
    REQUIRE_TRUE(fftSize > 0, 0, "whisper_mel_spectrogram: fftSize must be positive, got %d", fftSize);
    REQUIRE_TRUE(hopLength > 0, 0, "whisper_mel_spectrogram: hopLength must be positive, got %d", hopLength);
    REQUIRE_TRUE(numMelBins > 0, 0, "whisper_mel_spectrogram: numMelBins must be positive, got %d", numMelBins);
    REQUIRE_TRUE(targetFrames > 0, 0, "whisper_mel_spectrogram: targetFrames must be positive, got %d", targetFrames);

    helpers::whisperMelSpectrogram(block.launchContext(), input, sampleRate, fftSize, hopLength,
                                    numMelBins, targetFrames, lowerEdgeHz, upperEdgeHz, output);

    return sd::Status::OK;
}

DECLARE_SHAPE_FN(whisper_mel_spectrogram) {
    auto inputShapeInfo = inputShape->at(0);
    auto inputRank = shape::rank(inputShapeInfo);

    int numMelBins = block.numI() > 3 ? INT_ARG(3) : 80;
    int targetFrames = block.numI() > 4 ? INT_ARG(4) : 3000;

    sd::LongType batch = 1;
    if (inputRank == 2) {
        batch = shape::sizeAt(inputShapeInfo, static_cast<sd::LongType>(0));
    }

    // Output is always [batch, numMelBins, targetFrames] (or [numMelBins, targetFrames] for rank-1 input)
    std::vector<sd::LongType> outputShapeVec;
    if (inputRank == 2) {
        outputShapeVec = {batch, static_cast<sd::LongType>(numMelBins), static_cast<sd::LongType>(targetFrames)};
    } else {
        outputShapeVec = {static_cast<sd::LongType>(numMelBins), static_cast<sd::LongType>(targetFrames)};
    }

    return SHAPELIST(ConstantShapeHelper::getInstance().createShapeInfo(ArrayOptions::dataType(inputShapeInfo), 'c', outputShapeVec));
}

DECLARE_TYPES(whisper_mel_spectrogram) {
    getOpDescriptor()
        ->setAllowedInputTypes({ALL_FLOATS})
        ->setAllowedOutputTypes({ALL_FLOATS});
}

}  // namespace ops
}  // namespace sd

#endif
